File: 2021/day01/day01.py
#!/usr/bin/env python3
import sys

if len(sys.argv) > 1:
    filename = sys.argv[1]
else:
    filename = "input.txt"

data = [int(i) for i in open(filename).read().splitlines()]

# star 1
count = 0
for i, j in zip(data, data[1:]):
    if j > i:
        count += 1
print(count)

# star 2
windows = []
for i in range(len(data) - 2):
    windows.append(sum(data[i:i+3]))
count = 0
for i, j in zip(windows, windows[1:]):
    if j > i:
        count += 1
print(count)

File: 2021/day01/day01.hs
#!/usr/bin/env stack
-- stack --resolver lts-14.16 script

import System.Environment (getArgs)

pairs :: [a] -> [(a, a)]
pairs (x:xs) = zip (x:xs) xs
pairs _      = []

triples :: [a] -> [(a, a, a)]
triples (x:y:xs) = zip3 (x:y:xs) (y:xs) xs
triples _        = []

tripleSum :: (Int, Int, Int) -> Int
tripleSum (a, b, c) = a + b + c

star1 :: [Int] -> Int
star1 = sum . map (fromEnum . uncurry (<)) . pairs

star2 :: [Int] -> Int
star2 = star1 . map tripleSum . triples

main :: IO ()
main = do
  args <- getArgs
  let
    filename = case args of
      (x:_) -> x
      _     -> "input.txt"
  input <- map read . lines <$> readFile filename
  print $ star1 input
  print $ star2 input

